jxl/src/api.rs

```diff
@@ -54,36 +55,46 @@ pub trait JxlBitstreamInput {
 /// - `Complete(None)` if the prefix is definitively not a JXL signature.
 /// - `NeedsMoreInput` if the prefix matches a signature but is too short.
 pub fn check_signature(file_prefix: &[u8]) -> ProcessingResult<Option<JxlSignatureType>, ()> {
-    // Check for the longer, more specific container signature first.
-    if CONTAINER_SIGNATURE.starts_with(file_prefix) {
-        if file_prefix.len() >= CONTAINER_SIGNATURE.len() {
-            // We have the full container signature.
-            return ProcessingResult::Complete {
+    let prefix_len = file_prefix.len();
+
+    // Check for Container signature
+    let container_len = CONTAINER_SIGNATURE.len();
+    // Determine the number of bytes to compare (the length of the shorter slice)
+    let len_to_check = prefix_len.min(container_len);
+
+    // Compare the overlapping part of the prefix and the signature.
+    if file_prefix[..len_to_check] == CONTAINER_SIGNATURE[..len_to_check] {
+        // The prefix is a valid start. Now, is it complete?
+        return if prefix_len >= container_len {
+            // Yes, we have the full signature (or more).
+            ProcessingResult::Complete {
                 result: Some(JxlSignatureType::Container),
-            };
+            }
         } else {
-            // It's a valid prefix of the container signature, but we need more data.
-            return ProcessingResult::NeedsMoreInput {
-                size_hint: CONTAINER_SIGNATURE.len() - file_prefix.len(),
+            // No, we need more bytes to complete the signature.
+            ProcessingResult::NeedsMoreInput {
+                size_hint: container_len - prefix_len,
```

Comment:
I believe this is incorrect if `file_prefix` is [] :-) (or in general if we are in the common prefix of the two signatures, but that's empty)